### PR TITLE
Improve functional test

### PR DESCRIPTION
**Pull request by lucidguppy**
This splits out the functional tests I wrote into a separate folder so that travis ci doesn't fail.  It creates the server and then runs a couple of quick tests against it.  Eventually I'd like to add performance testing here too.

### CODE REVIEW

**Comment by sidorares** [comment]
looks good!


would it require less code without `Q` ? :) Also, what is the purpose of gulpfile? Could it be just a 2 lines in package.json scripts? "npm run test", "npm run functional-test"

File: functional_tests/test-server.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            .then(function(result){
+                assert.equal(result, a + b);
+                done();
+            })
+            .catch(function(err) {
+                throw new Error(err);
+            })
+            .done();
+    });
+});
```
Comment (sidorares) [inline functional_tests/test-server.js]: without Q, for example lines 91-101 would be

```js
launchService(bus, testName, testPath, testInterface, testImplementation);
bus.invoke(testInvocation, done);
```

**Comment by lucidguppy** [comment]
I tend to prefer using promises to avoid the callback pyramid.  I  suppose the test could be modified later to remove the need for Q and then the dependency dropped.

As for using npm functional test I would be fine with that - will update.

**Comment by lucidguppy** [comment]
I see your point about dropping q - let me try again and let you know.

**Comment by sidorares** [comment]
I understand your concern about pyramid, but mocha plays very nicely with standard node ( callback last, error first ) continuation passing style

Also you could move "start service" to `before` section, close bus in `after` and just do `if (process.CI) return;` at a very top level ( even not inside function ) if we assume no tests here would work in CI

**Comment by sidorares** [comment]
_if_ we decide to start using promises - take a look at https://github.com/domenic/chai-as-promised/

**Comment by lucidguppy** [comment]
Leaving the update at this point for the night.  I guess I understand callbacks better with promises than not.

**Comment by sidorares** [comment]
I'll make my version for you to look tomorrow but I'm ok with whatever works for you. It's just the less lines for the same functionality the better ( same about dependencies )

**Comment by lucidguppy** [comment]
Any word on this?  After this PR I'd like to work on some clean up.

**Comment by sidorares** [comment]
oh, sorry. I'll show you my suggestions within an hour

**Comment by sidorares** [comment]
can you look at https://github.com/sidorares/node-dbus/compare/improveFunctionalTest ?

**Comment by sidorares** [comment]
( haven't tried it yet, not on linux right now )

**Comment by lucidguppy** [comment]
How do you run the tests?  Tried npm test.  Tried node test-server.js.
Tried mocha.

**Comment by sidorares** [comment]
mocha test/functional/service.js

**Comment by lucidguppy** [comment]
I'll have to take a look at this tomorrow.  Trying to run it - I'm getting
this error.

module.js:340
    throw err;
          ^
Error: Cannot find module '../index.js'
